lib/rules/generic-type-naming.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * @fileoverview Enforces naming of generic type variables.
+ */
+"use strict";
+
+/**
+ * Check type parameters for rule match.
+ * @param {Array} typeParameters Nodes to be checked
+ * @param {any} context          ESLint context
+ * @param {Regex} regex          Rule name as regular expression
+ * @param {any} rule             Rule name as string
+ * @returns {void}
+ */
+function checkTypeParameters(typeParameters, context, regex, rule) {
+    const params = typeParameters && typeParameters.params;
+
+    if (!Array.isArray(params) || params.length === 0) {
+        return;
+    }
+    params.forEach(node => {
+        const type = node.type;
+
+        if (type === "TSTypeParameter" || type === "TypeParameter") {
+            const name = node.name;
+
+            if (name && !regex.test(name)) {
+                const data = { name, rule };
+
+                context.report({ node, messageId: "nameRule", data });
+            }
+        }
+    });
+}
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "Enforces naming of generic type variables",
+            category: "TypeScript",
+            url:
+                "https://github.com/nzakas/eslint-plugin-typescript/blob/master/docs/rules/generic-type-naming.md"
+        },
+        messages: {
+            nameRule: "Type parameter {{name}} does not match rule {{rule}}"
+        }
+    },
+
+    create(context) {
+        const rule = context.options[0];
+
+        if (!rule) {
+            return {};
+        }
+        const regex = new RegExp(rule);
+
+        return {
+            VariableDeclarator: node =>
+                checkTypeParameters(node.typeParameters, context, regex, rule),
```

Comment:
I wonder if you could use a higher-order function to DRY up this section:

```js
function createTypeParameterChecker(context, rule) {
  const regex = new RegExp(rule)
  return function checkTypeParameters(node) {
    const params = node.typeParameters && node.typeParameters.params
    // ... rest of logic you already have
  }
}
```

And then your rule would look something like this:

```js
create(context) {
  const rule = context.options[0]
  if (!rule) {
    return
  }

  const checkTypeParameters = createTypeParameterChecker(context, rule)
  return {
    VariableDeclarator: checkTypeParameters, // `node` gets passed to this function
    ClassDeclaration: checkTypeParameters,
    InterfaceDeclaration: checkTypeParameters,
    // etc.
  }
}
```

I like higher-order functions for situations like this, but I wonder if anyone else has feedback on this / if it makes sense. 😄